Sources/COWMacros/COWExpansionFactory.swift

```diff
@@ -218,30 +223,196 @@ internal class COWExpansionFactory<Context: MacroExpansionContext> {
   
   // MARK: Utilities
   
-  private func createDerivedStorageTypeDecl() -> StructDeclSyntax {
-    let members = appliedStructValidVarDecls.map {
-      MemberBlockItemSyntax(decl: $0)
+  // If `appliedStructDecl` conforms to Equatable and manually implements `==`,
+  // the derived storage must also conform to Equatable (otherwise we hit a
+  // compiler bug that breaks the build).
+  // https://forums.swift.org/t/macro-multiple-matching-functions-xyz-error-though-the-expanded-code-has-no-duplicate-functions-and-compiles/68700/4
+  // FIXME: remove the workaround when the compiler bug is fixed.
+  private func applyEquatableWorkaroundForDerivedStorage(
+    members: inout [MemberBlockItemSyntax],
+    protocols: inout [InheritedTypeSyntax]
+  ) {
+    // Bail out if derived storage doesn't conform to Equatable.
+    guard protocols.contains(where: {
+      guard let name = $0.type.identifier else {
+        return false
+      }
+      return equatableProtocolNames.contains(name) ||
+             hashableProtocolNames.contains(name)
+    }) else {
+      return
     }
     
+    // Bail out if `appliedStructDecl` does not manually implement `==`.
+    guard let nestingEqualFunc = appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap({$0.as(FunctionDeclSyntax.self)})
+      .first(where: {$0.likelyToConformToEquatable(for: appliedStructDecl)}) else {
+      return
+    }
+    
+    // Another compiler bug another workaround.
+    // The compiler doesn't call the manually-implemented `==` in `appliedStructDecl`;
+    // it will synthesize one for the derived storage and call it instead.
+    // Expand an `==` that is an exact copy (except for the signature) of
+    // `appliedStructDecl` to make sure the program behavior is correct.
+    
+    // In order to make this workaround work with computed properties and
+    // immutable functions with return values, we also have to copy them into
+    // the derived storage.
+    appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap { $0.as(VariableDeclSyntax.self) }
+      .filter { !$0.isStatic && !$0.isStored }
+      .forEach { members.append(MemberBlockItemSyntax(decl: $0)) }
+    appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap { $0.as(FunctionDeclSyntax.self) }
+      .filter { !$0.isStatic && !$0.isMutating && !$0.returnTypeEquals(to: "Void") }
+      .forEach { members.append(MemberBlockItemSyntax(decl: $0)) }
+    
+    guard let body = nestingEqualFunc.as(FunctionDeclSyntax.self)!.body else {
```

Comment:
Now I understood the mechanism of the "Equatable workaround" mentioned in this PR. However, why couldn't we just simply return `false` or `true` in the body? (instead of copying the @COW-applied struct's)

The derived storage hasn't to be `Equatable` when the @COW-applied struct is Equatable and has its own `static func ==` implemented. In this case, the return value of `static func ==` of the derived storage is an undefined. Thus the behavior of the `static func ==` of the derived storage hasn't to keep the same of the @COW-applied struct.

Reply:
I don't think you fully understand the purpose of the workaround. Please read the comment above again.

```swift
  // If `appliedStructDecl` conforms to Equatable and manually implements `==`,
  // the derived storage must also conform to Equatable (otherwise we hit a
  // compiler bug that breaks the build).
  // https://forums.swift.org/t/macro-multiple-matching-functions-xyz-error-though-the-expanded-code-has-no-duplicate-functions-and-compiles/68700/4
```

Because we have to implement `==` in the macro-generated struct, the control flow will go into it, so we must provide a correct implementation, rather than a pseudo one.

Reply:
I think the following screenshot could have a further explanation. In this case, the semantics of client code does not care about the behavior of `static func == (lhs: Self, rhs: Self) -> Bool` in the derived storage. We can just return `true` or `false`. This could simplify a lot of work, converging further issues bought by transferring members from @COW-applied structs.

<img width="817" alt="Screenshot 2024-04-20 at 8 47 19 PM" src="https://github.com/WeZZard/COWMacro/assets/960509/bc8a7bd9-61a5-4175-8957-b433368b6843">

Reply:
> In this case, the semantics of client code does not care about the behavior of static func == (lhs: Self, rhs: Self) -> Bool in the derived storage.

Hard NO. I already said that the control flow will enter the `==` in the derived storage, that's why we need a correct implementation. Please debug it yourself you are still confused about it.

Reply:
I found that I pasted a piece of unrelated comment yesterday, sorry about that. Since I finished this PR almost 2 months ago, I also forgot some of the details.

```swift
    // Another compiler bug another workaround.
    // The compiler doesn't call the manually-implemented `==` in `appliedStructDecl`;
    // it will synthesize one for the derived storage and call it instead.
    // Expand an `==` that is an exact copy (except for the signature) of
    // `appliedStructDecl` to make sure the program behavior is correct.
    
    // In order to make this workaround work with computed properties and
    // immutable functions with return values, we also have to copy them into
    // the derived storage.
```

Reply:
I've now investigated what happened in `testManuallyConformedEquatableStruct`. The syndrome is that:

1. For static dispatch of `static func ManuallyConformedEquatableStruct.==`, the manually implemented `==` is invoked.
2. For dynamic dispatch of `static func Equatable.==`, the compiler automatically generated `==` is invoked.

Demo code narrowed into case 1:

```swift
var fee = ManuallyConformedEquatableStruct()
let foe = fee
// Narrowed into case 1. No need to dispatch `static func ==` via ManuallyConformedEquatableStruct's protocol witness table of Equatable.
 let flag = fee == foe
```

Demo code narrowed into case 2:

```swift
var fee = ManuallyConformedEquatableStruct()
let foe = fee
// Narrowed into case 2, since XCTAssertEqual receives instances of a type that conforms to Equatable
XCTAssertEqual(fee, foe)
```

A further implication could be: the compiler auto generated an implementation for `static fun Equatable.==` for `ManuallyConformedEquatableStruct` and links it to the protocol witness table of `ManuallyConformedEquatableStruct : Equatable`. There are also evidence supports this implication: the types used in `ManuallyConformedEquatableStruct`'s members also are required to implement `Equatable`.

However, I still think the current solution to this compiler bug leaves no options when the automatic workaround fails. Instead, in my opinion, diagnosing this issue and generating template code to help developers work it around could be a more robust solution. I'm please to implement diagnosing the template code generating on top of this PR if you have no time to work on it.

Let's showcase the diagnose and fix process with the following 3 user stories:

## User Story 1: Adding the workaround without custom storage

I'm a `@COW` macro user.
  I would like the `@COW` macro diagnose the previous issue and generate a template code of the derived storage when I have no custom derived storage defined.
  Such that I can handle the derived storage's `static func ==` implementation by my self.

Step 1. `@COW` macro user has a struct with a custom `static func ==` implemented.

```swift
@COW
struct ManuallyConformedEquatableStruct: Equatable {
    
    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()
    
    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

Step 2: The `@COW` macro user waits the Xcode to highlight the code, the `@COW` macro runs in the background

```swift
@COW
// DIAGNOSTIC MESSAGE: Swift compiler cannot correct generate `static func Equatable.==` for this type.
struct ManuallyConformedEquatableStruct: Equatable {
    
    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()
    
    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

Step 3: The `@COW` macro user clicked the fix button of the diagnostic message

```swift
@COW
struct ManuallyConformedEquatableStruct: Equatable {
    
    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()

    // CODE ADDED BY THE FIX BUTTON BEGAN
    @COWStorage
    // `@COWStorageEquatableWorkaround` makes `MyStorage` to conform to `Equatable` and `COW.COWStorageEquatableWorkaround`
    @COWStorageEquatableWorkaround
    struct _$COWStorage: COW.CopyOnWriteStorage  {
    
        static func equalsWorkaround(lhs: Self, rhs: Self) -> Bool {
            <#implementation of static func ==#>
        }
    }
     // CODE ADDED BY THE FIX BUTTON ENDED

    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

Step 4: The `@COW` macro user fills <#implementation of static func ==#>

```swift
@COW
struct ManuallyConformedEquatableStruct: Equatable {
    
    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()

    // CODE ADDED BY THE FIX BUTTON BEGAN
    @COWStorage
    // `@COWStorageEquatableWorkaround` makes `MyStorage` to conform to `Equatable` and `COW.COWStorageEquatableWorkaround`
    @COWStorageEquatableWorkaround
    struct _$COWStorage: COW.CopyOnWriteStorage {
    
        static func equalsWorkaround(lhs: Self, rhs: Self) -> Bool {
            return lhs.wrappedValue == rhs.wrappedValue
        }
    }
     // CODE ADDED BY THE FIX BUTTON ENDED

    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

## User Story 2: Adding the workaround with custom storage

I'm a `@COW` macro user.
  I would like the `@COW` macro diagnose the previous issue and generate a template code of the `static func ==` for the derived storage when I already have custom derived storage defined.
  Such that I can handle the derived storage's `static func ==` implementation by my self.

Step 1. The `@COW` macro user has a struct with a custom `@COWStorage` and `static func ==` implemented.

```swift
@COW
struct ManuallyConformedEquatableStruct: Equatable {

    @COWStorage
    struct MyStorage  {
    
    }
    
    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()
    
    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

Step 2: The `@COW` macro user waits the Xcode to highlight the code. The `@COW` macro runs in the background

```swift
@COW
// DIAGNOSTIC MESSAGE: Swift compiler cannot correct generate `static func Equatable.==` for this type.
struct ManuallyConformedEquatableStruct: Equatable {

    @COWStorage
    struct MyStorage  {
    
    }
    
    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()
    
    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

Step 3: The `@COW` macro user clicked the fix button of the diagnostic message

```swift
@COW
struct ManuallyConformedEquatableStruct: Equatable {

    @COWStorage
    // `@COWStorageEquatableWorkaround` makes `MyStorage` to conform to `Equatable` and `COW.COWStorageEquatableWorkaround`
    // CODE ADDED BY THE FIX BUTTON BEGAN
    @COWStorageEquatableWorkaround
     // CODE ADDED BY THE FIX BUTTON ENDED
    struct MyStorage  {
    
        // CODE ADDED BY THE FIX BUTTON BEGAN
        static func equalsWorkaround(lhs: Self, rhs: Self) -> Bool {
            <#implementation of static func ==#>
        }
         // CODE ADDED BY THE FIX BUTTON ENDED
    }
    
    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()

    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

Step 4: The `@COW` macro user fills <#implementation of static func ==#>

```swift
@COW
struct ManuallyConformedEquatableStruct: Equatable {
    
    @COWStorage
    // `@COWStorageEquatableWorkaround` makes `MyStorage` to conform to `Equatable` and `COW.COWStorageEquatableWorkaround`
    // CODE ADDED BY THE FIX BUTTON BEGAN
    @COWStorageEquatableWorkaround
     // CODE ADDED BY THE FIX BUTTON ENDED
    struct MyStorage  {
    
        // CODE ADDED BY THE FIX BUTTON BEGAN
        static func equalsWorkaround(lhs: Self, rhs: Self) -> Bool {
            return lhs.wrappedValue == rhs.wrappedValue
        }
         // CODE ADDED BY THE FIX BUTTON ENDED
    }

    var value: Int = 0
    var wrappedValue: Int { value }
    let foo = NotConformingToHashable()

    static func == (lhs: Self, rhs: Self) -> Bool {
      return lhs.wrappedValue == rhs.wrappedValue
    }
    
  }
```

## User Story 3: Removing the workaround

I'm a `@COW` macro user.
  I would like the `@COW` macro to stop generating `@COWStorageEquatableWorkaround` for the derived storage when the previous compiler bug is fixed and the code is compiled against the latest compiler.
  Such that I can reduce the cost that maintains the code by just removing the `equalsWorkaround` in the derived storage.

Reply:
1. I don't think "shifting the burden of working around compiler bugs to the macro adopter" is acceptable from the user standpoint. I believe that dynamic dispatch of `Equatable` is a much more frequently used scenario, which makes this workaround basically a must for `struct`s manually conforming `Equatable` while containing non-`Equatable`-conforming types, which makes the seemingly opt-in of your design essentially a mandate.

2. Yes, I agree that adopter has little or even nothing to do when the current workaround fails. I'm OK to make this failure more user-friendly by pre-checking non-stored property usages, or even to try a completely different solution, like maintaining an unowned pointer to the outer `struct` and forwarding `==` calls to it - we are mostly likely to move to `class`-based solution later on so this should be safe and doable. Actually this was the first idea I came up with, but I haven't given it a try due to current `struct`-`class`-`struct` design.

Anyway, letting the user manually apply the workaround is the last resort imo.

Reply:
Forget the unowned pointer idea, it is not doable because we have no idea when a `struct` is getting released in Swift world, not to mention that the `class` instance will be shared among copies.

Here is a new idea I just come up with: what if we construct 2 temporary `struct`s on stack with the same storage instance and perform a static `==` in `equalsWorkaround`? What does this idea sound to you?